File: yongshub-python/대기업_계열사_코테/BOJ_17484.py
import sys

input = sys.stdin.readline

N, M = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(N)]
MAX_VAL = 100 * 1000 + 1
dp = [[[MAX_VAL]*3 for _ in range(M)] for _ in range(N)]
for y in range(N):
    if y == 0:
        for x in range(M):
            for d in range(3):
                dp[y][x][d] = board[y][x]
    else:
        for x in range(M):
            if x == 0:
                dp[y][x][0] = min(dp[y-1][x+1][1], dp[y-1][x+1][2]) + board[y][x]
                dp[y][x][1] = dp[y-1][x][0] + board[y][x]
            elif x == M-1:
                dp[y][x][1] = dp[y-1][x][2] + board[y][x]
                dp[y][x][2] = min(dp[y-1][x-1][0], dp[y-1][x-1][1]) + board[y][x]
            else:
                dp[y][x][0] = min(dp[y-1][x+1][1], dp[y-1][x+1][2]) + board[y][x]
                dp[y][x][1] = min(dp[y-1][x][0], dp[y-1][x][2]) + board[y][x]
                dp[y][x][2] = min(dp[y-1][x-1][0], dp[y-1][x-1][1]) + board[y][x]
answer = 1e9
for x in range(M):
    answer = min(min(dp[N-1][x]), answer)
print(answer)

File: lcomment-java/src/BOJ/DynamicProgramming/BOJ_17484.java
package BOJ.DynamicProgramming;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class BOJ_17484 {
    static int N, M;
    static int[][] map;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int[] input = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

        N = input[0];
        M = input[1];
        map = new int[N][M];
        // 0: 좌, 1: 직, 2: 우
        int[][][] dp = new int[N][M][3];

        for(int i=0 ; i<N ; i++){
            map[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

            /* 초기값 세팅
             * 맨 왼쪽은 오른쪽에서 올 수 없음
             * 맨 오른쪽은 왼쪽에서 올 수 없음
             */
            for(int j=0; j<M ; j++) Arrays.fill(dp[0][j], map[0][j]);
            dp[i][0][2] = Integer.MAX_VALUE;
            dp[i][M-1][0] = Integer.MAX_VALUE;
        }

        for(int i=1 ; i<N ; i++){
            for(int j=0 ; j<M ; j++){
                if(in(i-1, j-1)) dp[i][j][2] = Math.min(dp[i-1][j-1][0], dp[i-1][j-1][1]) + map[i][j];

                if(in(i-1, j)) dp[i][j][1] = Math.min(dp[i-1][j][0], dp[i-1][j][2]) + map[i][j];

                if(in(i-1, j+1)) dp[i][j][0] = Math.min(dp[i-1][j+1][1], dp[i-1][j+1][2]) + map[i][j];
            }
        }
        int min = Integer.MAX_VALUE;

        for(int i=0 ; i<M ; i++){
            min = Math.min(min, Math.min(dp[N-1][i][0], Math.min(dp[N-1][i][1], dp[N-1][i][2])));
        }

        System.out.println(min);
    }
    static boolean in(int r, int c){
        return 0<=r && r<=N && 0<=c && c<M;
    }
}
